kernel/kernel-api/src/main/java/io/delta/kernel/internal/skipping/StatsSchemaHelper.java

```diff
@@ -0,0 +1,304 @@
+                fields.add(new StructField(
+                    getPhysicalName(field), LongType.LONG, true));
+            }
+        }
+        return new StructType(fields);
+    }
+
+    //////////////////////////////////////////////////////////////////////////////////
+    // Private class helpers
+    //////////////////////////////////////////////////////////////////////////////////
+
+    /**
+     * Given a logical column and a stats type returns the corresponding column in the statistics
+     * schema
+     */
+    private Column getStatsColumn(Column column, String statType) {
+        checkArgument(logicalToPhysicalColumn.containsKey(column),
+            String.format("%s is not a valid leaf column for data schema", column, dataSchema));
+        return getChildColumn(logicalToPhysicalColumn.get(column), statType);
+    }
+
+    /**
+     * Given a data schema returns a map of {logical column -> (physical column, data type)} for
+     * all leaf columns in the schema.
+     */
+    private Map<Column, Tuple2<Column, DataType>> getLogicalToPhysicalColumnAndDataType(
+            StructType dataSchema) {
+        Map<Column, Tuple2<Column, DataType>> result = new HashMap<>();
+        for (StructField field : dataSchema.fields()) {
+            if (field.getDataType() instanceof StructType) {
+                Map<Column, Tuple2<Column, DataType>> nestedCols =
+                    getLogicalToPhysicalColumnAndDataType((StructType) field.getDataType());
+                for (Column childLogicalCol : nestedCols.keySet()) {
+                    Tuple2<Column, DataType> childCol = nestedCols.get(childLogicalCol);
+                    Column childPhysicalCol = childCol._1;
+                    DataType childColDataType = childCol._2;
+                    result.put(
+                        getChildColumn(childLogicalCol, field.getName()),
+                        new Tuple2<>(
+                            getChildColumn(childPhysicalCol, getPhysicalName(field)),
+                            childColDataType
+                        ));
+                }
+            } else {
+                result.put(
+                    new Column(field.getName()),
+                    new Tuple2<>(
+                        new Column(getPhysicalName(field)),
+                        field.getDataType()
+                    )
+                );
+            }
+        }
+        return result;
+    }
+
+    /** Returns the provided column as a child column nested under {@code parentName} */
+    private static Column getChildColumn(Column column, String parentName) {
+        return new Column(prependArray(column.getNames(), parentName));
+    }
+
+    /**
+     * Given an array {@code names} and a string element {@code preElem} return a new array with
+     * {@code preElem} inserted at the beginning
+     */
+    private static String[] prependArray(String[] arr, String preElem) {
+        String[] newNames = new String[arr.length + 1];
```

Comment:
how often will this be used? is this super common? it seems like we have chosen a data structure (array) that isn't very good for this use case (prepending). a LinkedList would be better. What are the other use cases of this array inside of column? Or, why was an array [] chosen?

Reply:
@vkorukanti what do you think? can we change this to be a List?
I think the `names` field is mostly manipulated (prepended or appended to) as part of the data skipping

Reply:
We can get rid of this if we follow the method I described in my other comment above about how to construction the leaf level columns. 

But I am also ok to make this a list. It is just that lists have extra baggage.

Reply:
Not sure if I understand, but what I am thinking is maintain the current root path and whenever we visit a child allocate n+1 array (system.cop can do it) and assign the child name at the `n` index. See the comment [here](https://github.com/delta-io/delta/pull/2433/files#r1443554254). In any case, I am ok to change it to List and if that becomes a problem we can convert it back to array.